### ISSUE TITLE
assertion failed when deserializing an Arena built with Arena::from_iter

### ISSUE THREAD

**Issue by Athorus** [fitzgen/generational-arena#27]
The following code fails, but it shouldn't.

When the arena is built by calling Arena::new and insert, everything is ok.

``` rust
use std::iter::FromIterator;
use generational_arena::Arena;
use serde::{Serialize, Deserialize};

#[derive(Serialize, Deserialize)]
pub struct Struct {
    arena: Arena<usize>
}

pub fn main() {
    let mut vec = vec![0];
    let arena = Arena::from_iter(vec.drain(..));
    let struc = Struct { arena };
    let ser = serde_yaml::to_string(&struc).unwrap();
    let struc: Struct = serde_yaml::from_str(&ser).unwrap();
}
```

**Comment by fitzgen** [comment]
Thanks for filing a bug! Would you like to investigate a fix?

**Comment by dmitmel** [comment]
This bug actually has nothing to do with `from_iter`. It happens when the vector `items` in the function `visit_seq`, used for storing arena entries, has some allocated capacity left, in other words when `items.capacity()` is greater than `items.len()` after consuming every element of the sequence. Note that it happens only when using data formats which don't store the number of items in an array (such as JSON and YAML), meaning that the function [`SeqAccess::size_hint`](https://docs.rs/serde/1.0.116/serde/de/trait.SeqAccess.html#method.size_hint) returns `None`, because in other cases (e.g. when using bincode) the `items` vector is pre-allocated with precisely the required capacity.

The assertion mentioned in the title of this issue fails because the logic for filling the rest of the capacity of `items` with `Empty` entries is incorrect: it doesn't fill the last available slot. I actually doubt that it is even useful: if you serialize an arena with 100 elements into JSON and deserialize it back - you get an arena of 128 elements, which after a second round of serialization-deserialization won't be the same as the initial arena. Can `Arena` work if the underlying vector is not filled entirely?

The fix in commit 219a4b0edaf9593345c9a96c23547e961967078e addresses the issue perfectly, but I don't think it deserves its own test or can be tested at all. However, I also wonder why the `+ 1` was there in the first place.

**Comment by dmitmel** [comment]
Oh, I see. The length really **must** be the same as capacity. In which case, perhaps instead of extending the arena with new entries `shrink_to_fit` can be called instead?
